### PR TITLE
Support write operations

### PR DESCRIPTION
**Pull request by antangelo**
Adds support for mkdir, rmdir, mknod, rename, truncate, write, and utimens FUSE operations. Tested an edited hard drive image to be working with xemu, but I don't have the ability to test on a real Xbox right now.

### CODE REVIEW

**Comment by mborgerson** [comment]
> Adds support for mkdir, rmdir, mknod, rename, truncate, write, and utimens FUSE operations. Tested an edited hard drive image to be working with xemu, but I don't have the ability to test on a real Xbox right now.

Wow, thanks for taking initiative on this! I'm very excited about your patch. Left a few comments, mostly nits. Please update the README to indicate that write support works.

File: src/fatx_file.c [hunk 1]
```diff
@@ -166,11 +191,242 @@ int fatx_read(struct fatx_fs *fs, char const *path, off_t offset, size_t size, v
...
+        }
+
+        if (size - total_bytes_written == 0)
+        {
+            fatx_debug(fs, "finished writing\n");
+            break;
+        }
+
+        /* Move to next cluster? */
+        fatx_error(fs, "cluster offset = %zx\n", cluster_offset);
```
[the remark this answers is not in the document]
Comment (antangelo) [reply src/fatx_file.c]: Fixed

File: src/fatx_file.c [hunk 2]
```diff
@@ -166,11 +191,242 @@ int fatx_read(struct fatx_fs *fs, char const *path, off_t offset, size_t size, v
...
+    /* Get file attributes. */
+    status = fatx_get_attr(fs, path, &attr);
+    if (status) return status;
+
+    size_t enc_clusters = 1;
+    size_t cluster = attr.first_cluster;
+    while (enc_clusters * fs->bytes_per_cluster < offset)
+    {
+        status = fatx_get_next_cluster(fs, &cluster);
+        if (status == FATX_STATUS_ERROR) // Out of clusters, alloc more
```
[the remark this answers is not in the document]
Comment (antangelo) [reply src/fatx_file.c]: Fixed

File: src/fatx_file.c [hunk 3]
```diff
@@ -166,11 +191,242 @@ int fatx_read(struct fatx_fs *fs, char const *path, off_t offset, size_t size, v
...
+int fatx_rename(struct fatx_fs *fs, char const *from, char const *to)
+{
+    fatx_debug(fs, "fatx_rename(from=\"%s\", to=\"%s\")\n", from, to);
+
+    struct fatx_attr attr;
+    char from_path[FATX_MAX_FILENAME_LEN+1], to_path[FATX_MAX_FILENAME_LEN+1];
+    int status;
+
+    /* Sanity check that we're not trying to move the file */
+    strcpy(from_path, from);
```
[the remark this answers is not in the document]
Comment (antangelo) [reply src/fatx_file.c:416]: Fixed, and added some more general checks in mknod/mkdir as well.

File: src/fatxfs_fuse.c [hunk 4]
```diff
@@ -378,7 +404,40 @@ int fatx_fuse_mkdir(const char *path, mode_t mode)
 
     status = fatx_mkdir(pd->fs, path);
 
-    return status;
+    switch (status)
```
[the remark this answers is not in the document]
Comment (antangelo) [reply src/fatxfs_fuse.c]: Fixed.

**Pull request merged**